src/inmanta/server/services/orchestrationservice.py

```diff
@@ -59,6 +62,161 @@
+        assert self.paired_resources is not None
+        paired_resources: List[PairedResource] = self.paired_resources
+        updated_resource_sets: Set[str] = set(
+            res.new_resource.resource_set for res in paired_resources if res.new_resource.resource_set
+        )
+
+        def copy_with_incremented_version(resource: Dict[str, Any]) -> Dict[str, Any]:
+            res = Id.parse_id(resource["id"])
+            res.increment_version()
+            resource["id"] = res.resource_version_str()
+            return resource
+
+        to_keep: List[Dict[str, Any]] = [
+            copy_with_incremented_version(r.resource)
+            for r in list(old_resources.values())
+            if r.resource_set not in self.removed_resource_sets
+            and (r.is_shared_resource() or r.resource_set not in updated_resource_sets)
+        ]
+
+        merged_resources: Dict[ResourceIdStr, Dict[str, Any]] = {r["id"]: r for r in to_keep}
+
+        for paired_resource in paired_resources:
+            new_resource = paired_resource.new_resource
+            old_resource = paired_resource.old_resource
+            assert new_resource is not None
+            assert (
+                old_resource is None
+                or old_resource is not None
+                and Id.parse_id(old_resource.resource["id"]).resource_str()
+                == Id.parse_id(new_resource.resource["id"]).resource_str()
+            )
+            if paired_resource.is_new_resource():
+                merged_resources[new_resource.resource["id"]] = new_resource.resource
+            else:
+                if paired_resource.resource_changed_resource_set():
+                    raise BadRequest(
+                        f"A partial compile cannot migrate a resource({new_resource.resource['id']}) " "to another resource set"
+                    )
+                if new_resource.resource_set is None and paired_resource.is_update():
+                    raise BadRequest(
+                        f"Resource ({new_resource.resource['id']}) without a resource set cannot"
+                        " be updated via a partial compile"
+                    )
+                else:
+                    merged_resources[new_resource.resource["id"]] = new_resource.resource
+        return list(merged_resources.values())
+
+    def _merge_resource_sets(self, old_resource_sets: Dict[ResourceIdStr, Optional[str]]) -> Dict[ResourceIdStr, Optional[str]]:
+        assert self.paired_resources is not None
+        paired_resources: List[PairedResource] = self.paired_resources
+        updated_resource_sets: Set[str] = set(
+            res.new_resource.resource_set for res in paired_resources if res.new_resource.resource_set
+        )
+        to_keep: List[str] = list(updated_resource_sets) + self.removed_resource_sets
+        new_resource_sets: Dict[ResourceIdStr, Optional[str]] = {k: v for k, v in old_resource_sets.items() if v not in to_keep}
```

Comment:
at that point it does indeed but the next call
`new_resource_sets.update(self.resource_sets)
`
adds the updates to new_resource_sets. Should I name this otherwise? (`update` returns nothing)

Reply:
I added a suggestion in a new comment to improve the naming.